File: Problems & Solutions/F/Find patterns/14 November 2022/Python/Find patterns.py
#User function Template for python3
class Solution:
    def numberOfSubsequences (self,S,W):
        ans=int()
        s=[char for char in S]
        w=[char for char in W]
        i,j=[0]*2
        while i<len(s) and j<len(w):
            if s[i]==w[j]:
                s[i]='dinesh'
                i+=1
                j+=1
            else:i+=1
            if len(w)==j:ans+=1;i,j=[0]*2
        return ans;

#{ 
 # Driver Code Starts
#Initial Template for Python 3
if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        
        S=str(input())
        W=str(input())

        ob = Solution()
        print(ob.numberOfSubsequences(S,W))

# } Driver Code Ends

File: Problems & Solutions/F/Find patterns/14 November 2022/Java/Find patterns.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class GFG
{
    public static void main(String args[])throws IOException
    {
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while(t-- > 0)
        {
            String S = read.readLine();
            String W = read.readLine();

            Solution ob = new Solution();
            System.out.println(ob.numberOfSubsequences(S,W));
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution{
    static int numberOfSubsequences(String S, String W){
        // code here
        int n=0,k=0,ans=0;
        char arr[]=new char[W.length()];
        for(int i=0;i<W.length();i++){
            arr[i]=W.charAt(i);
        }
        boolean f[]=new boolean[S.length()];
        for(int i=0;i<S.length();i++){
            if(f[i]==true)
            continue;
            if(f[i]==false && S.charAt(i)==arr[k]){
                k++;
                f[i]=true;
            }
            if(k==arr.length){
                i=0;
                k=0;
                ans++;
            }
        }
        return ans;
    }
}